be/src/storage/vectorized/compaction.cpp

```diff
@@ -48,27 +104,59 @@ Status Compaction::do_compaction_impl() {
 
     // 1. prepare input and output parameters
     int64_t segments_num = 0;
+    int64_t total_row_size = 0;
     for (auto& rowset : _input_rowsets) {
         _input_rowsets_size += rowset->data_disk_size();
         _input_row_num += rowset->num_rows();
         segments_num += rowset->num_segments();
+        total_row_size += rowset->total_row_size();
     }
+
     TRACE_COUNTER_INCREMENT("input_rowsets_data_size", _input_rowsets_size);
     TRACE_COUNTER_INCREMENT("input_row_num", _input_row_num);
     TRACE_COUNTER_INCREMENT("input_segments_num", segments_num);
 
     _output_version = Version(_input_rowsets.front()->start_version(), _input_rowsets.back()->end_version());
 
+    // choose vertical or horizontal compaction algorithm
+    auto iterator_num_res = _get_segment_iterator_num();
+    if (!iterator_num_res.ok()) {
+        LOG(WARNING) << "fail to get segment iterator num. tablet=" << _tablet->full_name()
+                     << ", err=" << iterator_num_res.status().to_string();
+        return iterator_num_res.status();
+    }
+    size_t segment_iterator_num = iterator_num_res.value();
+    int64_t max_columns_per_group = config::vertical_compaction_max_columns_per_group;
+    size_t num_columns = _tablet->num_columns();
+    CompactionAlgorithm algorithm =
+            choose_compaction_algorithm(num_columns, max_columns_per_group, segment_iterator_num);
+    if (algorithm == VERTICAL) {
+        // split columns into column group
+        get_column_groups(_tablet->num_columns(), _tablet->num_key_columns(), max_columns_per_group, &_column_groups);
+    }
+
+    // max rows per segment
+    int64_t max_rows_per_segment =
+            get_segment_max_rows(config::max_segment_file_size, _input_row_num, _input_rowsets_size);
+
     LOG(INFO) << "start " << compaction_name() << ". tablet=" << _tablet->full_name()
-              << ", output version is=" << _output_version.first << "-" << _output_version.second;
+              << ", output version is=" << _output_version.first << "-" << _output_version.second
+              << ", max rows per segment=" << max_rows_per_segment << ", segment iterator num=" << segment_iterator_num
+              << ", algorithm=" << compaction_algorithm_to_string(algorithm)
+              << ", column group size=" << _column_groups.size() << ", columns per group=" << max_columns_per_group;
 
-    RETURN_IF_ERROR(construct_output_rowset_writer());
+    // create rowset writer
+    RETURN_IF_ERROR(construct_output_rowset_writer(max_rows_per_segment, algorithm));
     TRACE("prepare finished");
 
     // 2. write combined rows to output rowset
     Statistics stats;
-    auto res = merge_rowsets(config::compaction_memory_limit_per_worker, &stats);
-
+    Status res;
+    if (algorithm == VERTICAL) {
+        res = merge_rowsets_vertical(&stats);
```

Comment:
merge_rowsets_vertically and merge_rowsets_horizontally may be a better name.

Reply:
ok